### PR TITLE
validate currency codes in PaymentRequest ctor

### PR DESCRIPTION
**Pull request by marcoscaceres**

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/259637289)

# Firefox (nightly) #

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/259637290)

# Sauce (safari) #

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/259637296)

# Chrome (unstable) #

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/259637297)

# Sauce (MicrosoftEdge) #

**Review by alphan102 — APPROVED** [review]
LGTM

**Comment by marcoscaceres** [comment]
@sideshowbarker, I'm hitting "stability" fail on a Travis CI job, but I'm not sure what it's asking me to fix. Any guidance would be greatly appreciated.

**Comment by sideshowbarker** [comment]
> @sideshowbarker, I'm hitting "stability" fail on a Travis CI job, but I'm not sure what it's asking me to fix. Any guidance would be greatly appreciated.

Generally the thing I do first is just to restart the Travis jobs to see if the cause was just some temporary failure.

But I did that and two of the jobs are still failing. I can’t personally make time to investigate it today but anyway @jgraham has a lot more insight and may be able to identify the cause just from a quick look. @gsnedders too

**Comment by sideshowbarker** [comment]
The failing jobs are:

* https://travis-ci.org/w3c/web-platform-tests/jobs/259637289
* https://travis-ci.org/w3c/web-platform-tests/jobs/259637290

**Comment by marcoscaceres** [comment]
@jgraham, @gsnedders, halp? I can't figure out what it's complaining about :( The exceptions there are expected because no browser supports throwing the `RangeError` yet... which, iiuc, that's the error.

File: payment-request/payment-request-ctor-currency-code-checks.https.html [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    const details = {
+      total: defaultTotal,
+      modifiers: [modifier],
+    };
+    assert_throws(
+      RANGE_ERROR,
+      () => {
+        new PaymentRequest(defaultMethods, details);
+      },
+      `Expected RangeError with invalid shippingOption currency code "${invalidCurrency}".`
```
Comment (bentian) [inline payment-request/payment-request-ctor-currency-code-checks.https.html]: s/shippingOption/modifier

File: payment-request/payment-request-ctor-currency-code-checks.https.html [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    };
+    try {
+      new PaymentRequest(defaultMethods, details);
+    } catch (err) {
+      assert_unreached(
+        `Unexpected error with valid additionalDisplayItems[n] currency code "${validCurrency}": ${err.message}`
+      );
+    }
+  }
+}, "Check and canonicalize valid modifiers[n].total amount.");
```
Comment (bentian) [inline payment-request/payment-request-ctor-currency-code-checks.https.html]: s/total/additionaDisplayItem

File: payment-request/payment-request-ctor-currency-code-checks.https.html [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+      RANGE_ERROR,
+      () => {
+        new PaymentRequest(defaultMethods, details);
+      },
+      `Expected RangeError with invalid shippingOption currency code "${invalidCurrency}".`
+    );
+  }
+}, "Check and canonicalize valid modifiers[n].total amount and rethrow RangeError.");
+
+test(() => {
```
Comment (bentian) [inline payment-request/payment-request-ctor-currency-code-checks.https.html]: Remove this test duplicate to [above](https://github.com/w3c/web-platform-tests/pull/6670/files#diff-b4079994269456941761c85b811c258fR215).

File: payment-request/payment-request-ctor-currency-code-checks.https.html [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    };
+    assert_throws(
+      RANGE_ERROR,
+      () => {
+        new PaymentRequest(defaultMethods, details);
+      },
+      `Expected RangeError with invalid shippingOption currency code "${invalidCurrency}".`
+    );
+  }
+}, "Check and canonicalize valid details.shippingOptions amount and rethrow RangeError.");
```
Comment (bentian) [inline payment-request/payment-request-ctor-currency-code-checks.https.html]: s/valid/invalid

File: payment-request/payment-request-ctor-currency-code-checks.https.html [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+    };
+    assert_throws(
+      RANGE_ERROR,
+      () => {
+        new PaymentRequest(defaultMethods, details);
+      },
+      `Expected RangeError with invalid displayItem currency "${invalidCurrency}".`
+    );
+  }
+}, "Check and canonicalize valid details.displayItems amount and rethrow RangeError.");
```
Comment (bentian) [inline payment-request/payment-request-ctor-currency-code-checks.https.html]: s/valid/invalid

File: payment-request/payment-request-ctor-currency-code-checks.https.html [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+    };
+    assert_throws(
+      RANGE_ERROR,
+      () => {
+        new PaymentRequest(defaultMethods, details);
+      },
+      `Expected RangeError with invalid shippingOption currency code "${invalidCurrency}".`
+    );
+  }
+}, "Check and canonicalize valid modifiers[n].total amount and rethrow RangeError.");
```
Comment (bentian) [inline payment-request/payment-request-ctor-currency-code-checks.https.html]: s/valid/invalid

File: payment-request/payment-request-ctor-currency-code-checks.https.html [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+    };
+    assert_throws(
+      RANGE_ERROR,
+      () => {
+        new PaymentRequest(defaultMethods, details);
+      },
+      `Expected RangeError with invalid additionalDisplayItems[n] currency code "${invalidCurrency}".`
+    );
+  }
+}, "Check and canonicalize invalid modifiers[n].total amount and rethrow RangeError.");
```
Comment (bentian) [inline payment-request/payment-request-ctor-currency-code-checks.https.html]: s/total/additionalDisplayItems

**Comment by gsnedders** [comment]
@bobholt I suspect you don't care that the old tool didn't post comments here, but it didn't.

@marcoscaceres the problem is that it's getting 20 results for one subtest, having run the test 10 times: that implies that one of the test names isn't unique, specifically `/payment-request/payment-request-ctor-currency-code-checks.https.html` | `Check and canonicalize valid modifiers[n].total amount.`

**Comment by bobholt** [comment]
@gsnedders. We just deployed a fix for this to prbuildbot this morning, but are also _very close_ to moving to the new bot.

**Comment by marcoscaceres** [comment]
@gsnedders, thank you! Yeah, that was totally not clear from the error message. 

@bobholt, looking forward to seeing what the new bot reports... if you all need someone external to evaluate the comprehensibility of error messages, please just give me a shout. I'm all for making error messages more "human" - and provide tips on how to fix things in those error messages.

File: payment-request/payment-request-ctor-currency-code-checks.https.html (hunk 3 above)
Comment (marcoscaceres) [inline payment-request/payment-request-ctor-currency-code-checks.https.html]: Nice catch! Thank you.

**Pull request merged**